src/lustre/lustreAstDependencies.ml

```diff
@@ -20,112 +20,272 @@
+  ; nsummary = empty_node_summary
+  }
+  
+let add_pos: id_pos_map -> LA.ident -> Lib.position -> id_pos_map = fun m i p ->
+  IMap.update i
+    (function
+     | None -> Some ([p])
+     | Some ps -> Some (p :: ps))
+    m
+
+let union_pos: id_pos_map -> id_pos_map -> id_pos_map = fun m1 m2 ->
+  IMap.union (fun k v1 v2 -> Some (v1@v2)) m1 m2
+
+let singleton_pos: LA.ident -> Lib.position -> id_pos_map = fun i p ->
+  IMap.singleton i [p]
 
+let find_id_pos: id_pos_map -> LA.ident -> Lib.position option = fun m i ->
+  match (IMap.find_opt i m) with
+  | None -> None
+  | Some [] -> None
+  | Some (p::ps) -> Some p 
+                  
+let singleton_dependency_analysis_data: string -> LA.ident -> Lib.position -> dependency_analysis_data =
+  fun suffix i p -> { graph_data = G.singleton (suffix ^ i)
+                    ; id_pos_data = singleton_pos (suffix ^ i) p
+                    ; csummary  = empty_contract_summary
+                    ; nsummary = empty_node_summary } 
+
+let union_dependency_analysis_data : dependency_analysis_data -> dependency_analysis_data -> dependency_analysis_data = 
+  fun { graph_data = g1; id_pos_data = pos_m1; csummary = cs1; nsummary = ns1 }
+      { graph_data = g2; id_pos_data = pos_m2; csummary = cs2; nsummary = ns2 }
+  -> { graph_data = G.union g1 g2
+     ; id_pos_data = union_pos pos_m1 pos_m2
+     ; csummary = IMap.union (fun k v1 v2 -> Some v2) cs1 cs2
+     ; nsummary = IMap.union (fun k v1 v2 -> Some v2) ns1 ns2 }
+
+let connect_g_pos: dependency_analysis_data -> LA.ident -> Lib.position -> dependency_analysis_data =
+  fun ad i p -> { ad with graph_data = G.connect ad.graph_data i
+                        ; id_pos_data = add_pos ad.id_pos_data i p }  
+
+let remove: dependency_analysis_data -> LA.ident -> dependency_analysis_data =
+  fun ad i -> {ad with graph_data = G.remove_vertex ad.graph_data i}
+                      
+let map_g_pos: (LA.ident -> LA.ident) -> dependency_analysis_data -> dependency_analysis_data =
+  fun f ad -> 
+     let pos_info' = List.fold_left (fun m (k, v) -> IMap.add k v m) IMap.empty
+                       (List.map (fun (k, v) -> (f k, v)) (IMap.bindings ad.id_pos_data)) in
+     let g' = G.map f ad.graph_data in 
+     {ad with graph_data = g'; id_pos_data = pos_info'} 
+                      
+                      
 (* Suffixes for declaration types *)
 let ty_suffix = "type "
 let const_suffix = ""
 let node_suffix = ""
 let contract_suffix = "contract "
 let mode_suffix = "mode "
 
-let rec mk_graph_type: LA.lustre_type -> G.t = function
-  | TVar (_, i) -> G.singleton (ty_suffix ^ i)
+(*****************************************************************************
+ * Type 1: Dependency Analysis for top level type and constant declarations  *
+ *****************************************************************************)
+
+let rec drop_last: 'a list -> 'a list
+  = function
+  | [] -> failwith "drop_last"
+  | [e] -> []
+  | e :: r -> e :: drop_last r
```

Comment:
This function should go to `utils/lib.ml`

Reply:
done